pennylane_qiskit/qiskit_device2.py

```diff
@@ -53,6 +64,136 @@
 Result_or_ResultBatch = Union[Result, ResultBatch]
 
 
+def _track_execute(untracked_execute):
+    """Adds tracking to the execute method for Qiskit."""
+
+    @wraps(untracked_execute)
+    def execute(self, circuits, execution_config=DefaultExecutionConfig):
+        results = untracked_execute(self, circuits, execution_config)
+        if isinstance(circuits, QuantumScript):
+            batch = (circuits,)
+            batch_results = (results,)
+        else:
+            batch = circuits
+            batch_results = results
+        if self.tracker.active:
+            self.tracker.update(batches=1)
+            self.tracker.record()
+            for r, c in zip(batch_results, batch):
+                qpu_executions, shots = get_num_shots_and_executions(c)
+                if c.shots:
+                    self.tracker.update(
+                        simulations=1,
+                        executions=qpu_executions,
+                        results=r[0],
+                        shots=shots,
+                        resources=c.specs["resources"],
+                        errors=c.specs["errors"],
+                    )
+                else:
+                    self.tracker.update(
+                        simulations=1,
+                        executions=qpu_executions,
+                        results=r[0],
+                        resources=c.specs["resources"],
+                        errors=c.specs["errors"],
+                    )
+                self.tracker.record()
+        return results
+
+    return execute
+
+
+# pylint: disable=protected-access
+def qiskit_simulator_tracking(cls: type) -> type:
+    """Modifies all methods to add default simulator style tracking. Specifically for Qiskit devices.
+
+    Args:
+        cls (type): a subclass of :class:`pennylane.devices.Device`
+
+    Returns
+        type: The inputted class that has now been modified to update the tracker upon function calls.
+
+    Simulator style tracking updates:
+
+    * ``executions``: the number of unique circuits that would be required on quantum hardware
+    * ``shots``: the number of shots
+    * ``resources``: the :class:`~.resource.Resources` for the executed circuit.
+    * ``"errors"``: combined algorithmic errors from the quantum operations executed by the qnode.
```

Comment:
Is there a way to do some kind of wrapping for this instead? I literally experimented for like 3 hours today but it kept on running the original tracker function as well.

```
from pennylane.devices.modifiers import simulator_tracking

def custom_simulator_tracking(cls):
    """Decorator that adds custom tracking to the device class."""
    cls = simulator_tracking(cls)
    
    original_init = cls.__init__
    
    def new_init(self, *args, **kwargs):
        original_init(self, *args, **kwargs)
        self.execute = _track_execute(self.execute)
    
    cls.__init__ = new_init
    return cls
```

I tried something like this but it is bugged basically. My impression is that this should be doable without copy-paste bruteforce but I'm pretty unfamiliar with python wrappers/decorators so...